src/core/NPDtree.ml

```diff
@@ -151,24 +160,27 @@ module Make(E : Index.EQUATION) = struct
           Leaf.fold_match ~subst (Scoped.set dt trie.leaf) t k';
           Util.enter_prof prof_npdtree_retrieve;
         | Some i ->
-          match TC.view i.cur_term with
-            | (TC.Var _ | TC.DB _ | TC.AppBuiltin _ | TC.NonFO) ->
-              begin match trie.star with
-                | None -> ()
-                | Some subtrie ->
-                  traverse subtrie (next i)  (* match "*" against "*" *)
-              end
-            | TC.App (s, _) ->
-              begin try
-                  let subtrie = find_sub trie.map s in
-                  traverse subtrie (next i)
-                with NoSuchTrie -> ()
-              end;
-              begin match trie.star with
-                | None -> ()
-                | Some subtrie ->
-                  traverse subtrie (skip i)  (* skip subterm *)
-              end
+          if index_as_star i.cur_term
```

Comment:
In both cases you have a `if` followed by a partial function (`T.as_const_exn`). It would be even better if you could have:

```ocaml
type view_head =
  | As_star
  | As_app of ID.t * t list

let view_head (t:term) : view_head = … 
```

and you would `match view_head i.cur_term with … | …` in a safer way.

Reply:
I am trying to understand your commit policy. Should I amend this or put in a second commit?